pod/perlhacktips.pod

```diff
@@ -53,25 +53,101 @@ supported"> for further discussion about context.
+by a C compiler. Instead Perl's thread safe serialization, is done with an
+internal API with names like C<MUTEX_INIT()> and C<MUTEX_LOCK()> .
+
+Historically, atomic operations didn't exist on most CPU archs that Perl uses.
+If they existed, atomic APIs were always OS and vender specific, and never
+portable.
+
+As of 5.35.5, perl dropped support for a strict C89 compiler and moved to
+a minimum requirement of C89+some C99. See L</C99>. C11 standardized some
+atomics for the first time in the optionally implemented C<stdatomic.h>.
+Patches are welcome to add a portable atomic API, with fallbacks to
+C<MUTEX_LOCK()>.
+
+The right way to introduce a new C global variable, usually will be to add
+it as a new interpreter variable. See F<intrpvar.h>.  Since 5.10.0, adding
+or removing or changed the size of any interpreter variable, is not supported
+and undefined behavior.  Recompiling XS modules is required.
+
+There are some loopholes to this policy if you are writing unstable
+experiments.  These loopholes can never be used, in stable code, for the
+interpreter, or XS modules.  The loopholes may temporarily work, just long
+enough, to finish the experiment.  Remember, failure to get a C<SEGV>, or
+failure to get fatal C<panic:> error, doesn't mean you didn't introduce a bug,
+or corrupt a random malloc() block.
+
+Between 5.10.0, and upto 5.21.5, there was a provision, that adding 1 new
+variable at the end of F<intrpvar.h> as the very last member, was always binary
+compatible with older XS modules. This was intended only for stable
+maintenance releases. Ex, new maintenance release 5.18.1, loading an XS module
+compiled against header files from 5.18.0. Remember a newer 5.18.1 core,
+loading an XS binary compiled against 5.17.10 or 5.16.0, isn't allowed.
+
+So if cutting off current struct members in F<intrpvar.h>, didn't introduce a
+crash, you saved some time in your experiment and it was good luck.
+
+Starting with 5.21.6, stricter checking was added, to match the definition of
+F<intrpvar.h> as understood by each build of the perl interpreter binary or the
+C<libperl> binary, against the definition of F<intrpvar.h> as understood,
+when the XS module's shared library file was compiled.
+
+The exact sanity check requires struct length of C<PerlInterpreter *> aka
+C<my_perl> to be C<sizeof(PerlInterpreter)> or C<sizeof(*my_perl)>
+identical between Core and an XS module, regardless if its a non-threaded or
+threaded build of perl.  If the C compile time byte lengths don't match at
+runtime, L<perldiag/"%s: loadable library and perl binaries are mismatched (got %s handshake key %p, needed %p)">
+error happens.
+
+For 5.21.6 and up, to avoid recompiling XS, if you want to add a new interpreter
+global variable while hacking on the interpreter, is to rename, repurpose, or
+make into union, a current variable from F<intrpvar.h> without change its size,
+alignment, and offset.
```

Comment:
Preferably we'd fix this bug in the XS handshake, so that maint *can* add new members to the end of the interpreter structure.